### PR TITLE
Introduce MBMenuSurface

### PR DESCRIPTION
**Pull request by HannahKiekens**
Intended UseCase:

``` razor
    <MBIconButton @onclick="@OpenMenuSurfaceAsync" Icon="account_circle" />
    <div class="mdc-menu-surface--anchor">
        <MBMenuSurface @ref="MenuSurface" >
            <div>
                <div style="height:100px; width:100px; background-size: contain; background-image: url('_content/Material.Blazor.Website/icon-512.png');"></div>
                <div class="mdc-typography mdc-typography--headline6">
                            Hannah Kiekens
                </div>
                <div class="mdc-typography mdc-typography--subtitle1">
                            hannahkiekens@gmail.com
                </div>
            </div>
            <MBIconButton Icon="logout"/>
            <MBIconButton Icon="settings"/>
        </MBMenuSurface>
    </div>
@code {
    private MBMenuSurface MenuSurface { get; set; }

    private async Task OpenMenuSurfaceAsync()
    {
        await MenuSurface.ToggleAsync();
    }
}
```

### CODE REVIEW

**Comment by HannahKiekens** [comment]
Ok, what I really needed was an MBMenuSurface type component

I am looking for some feedback, if this looks good, I'll finish documentation for this Component

**Comment by MarkStega** [comment]
Hannah,

On first review this looks good. It would definitely be a 'plus' component since Material doesn't define such a component (except as part of a menu).

I have a couple of questions:

1) What makes it desirable to make this a component of Material.Blazor as opposed to a component in your application? I may be missing something but I didn't see anything in the implementation that made it 'better' by being part of M.B

2) Before documenting could you add at least a minimal demo page in Material.Blazor.Website? Just implementing your example from the beginning of this PR should be sufficient.

I am going to ask Simon to take a look at this PR also...

**Comment by simonziegler** [comment]
Hey @HannahKiekens, thanks for contributing this. @MarkStega and I just had a conversation and decided that I will be reviewing your PR. I have quite a lot of work on right now so anticipate getting there in perhaps a couple of weeks.

Simon

**Comment by HannahKiekens** [comment]
@MarkStega 
Having this Component in Material.Blazor would be a great extension point for custom Picker components & freeform menus, like the demo code. 
My first attempt involved modifying ComponentFoundation #419 so I could inherit from it in my own project. That did not seem like the correct solution.

**Comment by HannahKiekens** [comment]
@simonziegler 
Thx, I'll make sure to get this pull request up to standards. 


I am working on a major CMS/Dashboard rewrite. Improving extension points like this just seem best to do upstream.
Material.Blazor has been easy to work with and understand for my team.

**Comment by simonziegler** [comment]
That's great to hear, and I'd love to know more.

**Comment by MarkStega** [comment]
1) The demo page is looking good, but I am surprised to see a vertical scroll bar on the surface. Is that intentional?
2) I would think that on selection of either of the two menu items that the surface should close & you should follow the 'convention of menu.razor' and do the toast showing the selection. Do you agree?

With those two comments addressed and a documentation page I think this should be good to go.

**Comment by HannahKiekens** [comment]
This build-documentation failure doesnt seem related to what I did?

Critical failure uploading 'D:\a\Material.Blazor\Material.Blazor\siteDocFx\api\Material.Blazor.MBSlidingContent-1.html'

**Comment by stefanloerwald** [comment]
> This build-documentation failure doesnt seem related to what I did?
> 
> Critical failure uploading 'D:\a\Material.Blazor\Material.Blazor\siteDocFx\api\Material.Blazor.MBSlidingContent-1.html'

This looks like a spurious error. I will just trigger a re-run and see if it resolves. There were quite a few changes to the workflow  recently, so it might be a regression, but it most certainly doesn't look like an error in your code.

**Comment by MarkStega** [comment]
Hannah,

It's looking good; I gave you a two part question which was bad form (#2 above). You addressed the toast so hear is what should have been a separate question:

3) I would think that on selection of either of the two menu items that the surface should close. Do you agree? If not, what is the workflow that you envision for the surface?

**Comment by HannahKiekens** [comment]
@MarkStega 
> 1. I would think that on selection of either of the two menu items that the surface should close. Do you agree? If not, what is the workflow that you envision for the surface?

I intended this to be a dumb wrapper for @material/menu-surface.
The typescript API is ever so slightly different than a full menu

There are 2 usecases:

1. Freeform menus, like the example, wich is intended to wrap a C# ClaimsPrincipal. 
2. Custom Pickers, like ColorPicker ...

I believe there should not be any added intelligence here. This Component is intended to be used in other more complex Components, written in pure C#

**Comment by MarkStega** [comment]
Perhaps the best 'solution' then is to expand the demo a very small amount to show the surface in a more complete use case or two...

**Comment by simonziegler** [comment]
@HannahKiekens we just merged a fairly substantial PR of mine and next are going to merge one from @stefanloerwald in the coming few days. These will both change things a little for your code. First `DestroyMcwComponent` has gone, and second Stefan has changed attribute splatting quite dramatically. Hopefully the changes you will need to make are moderate. I suggest waiting for Stefan's PR to be merged before you update your code.

**Comment by stefanloerwald** [comment]
I'm going to perform a few more changes, but first will have some lunch. In the mean time we'll probably see a few broken builds, but that's fine.

File: Material.Blazor/Components/MenuSurface/MBMenuSurface.razor [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿@namespace Material.Blazor
+@inherits ComponentFoundation
+
+
+<div @ref="ElementReference"
+     @attributes="@AttributesToSplat()">
```
Comment (stefanloerwald) [inline Material.Blazor/Components/MenuSurface/MBMenuSurface.razor:9]: To be replaced by 
```
class="mdc-menu-surface mdc-menu-surface--fixed @ActiveConditionalClasses @(@class)"
style="@style"
id="@id"
@attributes="@AttributesToSplat()"
```
when #421 is merged.

File: Material.Blazor/Components/MenuSurface/MBMenuSurface.razor.cs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        private ElementReference ElementReference { get; set; }
+        private bool IsOpen { get; set; } = false;
+
+
+        // Would like to use <inheritdoc/> however DocFX cannot resolve to references outside Material.Blazor
+        protected override async Task OnInitializedAsync()
+        {
+            await base.OnInitializedAsync();
+
+            ClassMapperInstance
```
Comment (stefanloerwald) [inline Material.Blazor/Components/MenuSurface/MBMenuSurface.razor.cs]: To be replaced with `ConditionalCssClasses` when #421 is merged.

File: Material.Blazor/Components/MenuSurface/MBMenuSurface.razor.cs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                .Add("mdc-menu-surface mdc-menu-surface--fixed")
```
Comment (stefanloerwald) [inline Material.Blazor/Components/MenuSurface/MBMenuSurface.razor.cs]: To be removed when #421 is merged.

File: Material.Blazor/Components/MenuSurface/MBMenuSurface.razor.cs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        /// <summary>
+        /// For Material Theme to notify of menu closure via JS Interop.
+        /// </summary>
+        [JSInvokable("NotifyClosedAsync")]
+        //[System.Diagnostics.CodeAnalysis.SuppressMessage("Style", "IDE0060:Remove unused parameter", Justification = "<Pending>")]
+        public async Task NotifyClosedAsync()
+        {
+            IsOpen = false;
+            await Task.CompletedTask;
+        }
```
Comment (stefanloerwald) [inline Material.Blazor/Components/MenuSurface/MBMenuSurface.razor.cs]: Consider replacing with
```cs
[JSInvokable]
public void NotifyClosed()
{
   IsOpen = false;
}
```
I don't see a reason this call needs to be async at all. I also don't think the argument to JSInvokable is necessary unless the method name differs.

File: Material.Blazor/Components/MenuSurface/MBMenuSurface.razor (hunk 1 above)
Comment (stefanloerwald) [reply Material.Blazor/Components/MenuSurface/MBMenuSurface.razor:9]: Actually, `mdc-menu-surface--fixed` looks like it should be conditional, and seems to already be handled by the class mapper (soon to be named ConditionalCssClasses).

Comment (stefanloerwald) [reply Material.Blazor/Components/MenuSurface/MBMenuSurface.razor:9]: @HannahKiekens, please review whether `mdc-menu-surface--fixed` should be applied unconditionally or conditionally and just leave a comment here. I'll then adjust the code together with the changes necessary after #421 in one go.

File: Material.Blazor/Components/MenuSurface/MBMenuSurface.razor.cs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        /// <inheritdoc/>
+        private protected override async Task InstantiateMcwComponent() => await JsRuntime.InvokeVoidAsync("MaterialBlazor.MBMenuSurface.init", ElementReference, ObjectReference);
+
+
+        /// <summary>
+        /// Returns a menu surface class determined by the parameter.
+        /// </summary>
+        /// <param name="surfacePositioning"></param>
+        /// <returns></returns>
+        internal static string GetMenuSurfacePositioningClass(MBMenuSurfacePositioning surfacePositioning) =>
```
Comment (stefanloerwald) [inline Material.Blazor/Components/MenuSurface/MBMenuSurface.razor.cs]: Consider declaring this private. I can't see why this is necessary outside this component.

**Comment by simonziegler** [comment]
@HannahKiekens, we managed to get through what we needed to deal with so that we can now consider your PR. You'll notice that @stefanloerwald has updated your code a bit to catch up with the conflicts created by what we've been doing. Can you please have a look at what @MarkStega suggested about the website demo page and let us know when you feel ready for us to review the PR again? If you can manage this within the next few days we'll look to include this in the upcoming preview4.

**Review by simonziegler — CHANGES_REQUESTED** [review]
I'm fully happy with this PR except for the need to respond to @stefanloerwald's comments.

**Comment by stefanloerwald** [comment]
I just performed those minor changes myself, which also makes it consistent with the rest of the code base. Fine to merge from my side

**Pull request merged**